styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -50,32 +74,115 @@
   private final Time time;
   private final TimeoutConfig ttls;
   private final StateManager stateManager;
+  private final WorkflowCache workflowCache;
+  private final Storage storage;
 
-  public Scheduler(Time time, TimeoutConfig ttls, StateManager stateManager) {
+  public Scheduler(Time time, TimeoutConfig ttls, StateManager stateManager,
+                   WorkflowCache workflowCache, Storage storage) {
     this.time = Objects.requireNonNull(time);
     this.ttls = Objects.requireNonNull(ttls);
     this.stateManager = Objects.requireNonNull(stateManager);
+    this.workflowCache = Objects.requireNonNull(workflowCache);
+    this.storage = Objects.requireNonNull(storage);
   }
 
   void tick() {
-    for (Map.Entry<WorkflowInstance, RunState> entry : stateManager.activeStates().entrySet()) {
-      final WorkflowInstance key = entry.getKey();
-      final RunState state = entry.getValue();
+    final Map<String, Resource> resources;
+    try {
+      resources = storage.resources().stream().collect(toMap(Resource::id, identity()));
+    } catch (IOException e) {
+      LOG.warn("Failed to get resource limits", e);
+      return;
+    }
 
-      if (hasTimedOut(state)) {
-        LOG.info("Found stale state, issuing timeout for {}", state);
-        stateManager.receiveIgnoreClosed(Event.timeout(key));
-      }
+    final List<InstanceState> activeStates = stateManager.activeStates().entrySet().stream()
+        .map(entry -> InstanceState.create(entry.getKey(), entry.getValue()))
+        .collect(toList());
+
+    final Set<WorkflowInstance> timedOutInstances =
+        activeStates.parallelStream()
+            .filter(entry -> hasTimedOut(entry.runState()))
+            .map(InstanceState::workflowInstance)
+            .collect(toSet());
+
+    final Map<WorkflowId, Set<String>> workflowResourceReferences =
+        activeStates.parallelStream()
+            .map(InstanceState::workflowInstance)
+            .map(WorkflowInstance::workflowId)
+            .distinct()
+            .collect(toMap(workflowId -> workflowId, this::workflowResources));
+
+    final Map<String, Long> currentResourceUsage =
+        activeStates.parallelStream()
+            .filter(entry -> !timedOutInstances.contains(entry.workflowInstance()))
+            .filter(entry -> entry.runState().state() != RunState.State.QUEUED)
```

Comment:
After we have filtered out the instances about to time out, shouldn't we then filter only instances whose state is `RUNNING`?